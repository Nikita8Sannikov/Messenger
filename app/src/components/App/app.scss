@import "../../styles/variables.scss";
@import "../../styles/colors.scss";

.root {
    min-height: 100vh;
    width: 100%;
    background: $mainBackground;

    &__wrapper {
        width: inherit;
        display: flex;
        align-items: center;
        justify-content: center;
        padding-bottom: 20px;

        &__container {
            width: $contentWidth;
            display: flex;
            align-items: flex-start;
            justify-content: center;

            &__no-auth {
                width: 100%;
            }

            &__content {
                flex: 1;
                margin-top: $marginTop;
            }
        }
    }
}