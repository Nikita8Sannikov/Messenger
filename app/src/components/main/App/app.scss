.root {
	min-height: 100vh;
	width: 100%;
	background: $main-background;
	overflow-x: hidden;
	position: relative;

	&__wrapper {
		width: inherit;
		display: flex;
		align-items: center;
		justify-content: center;
		padding-bottom: 20px;

		&__container {
			width: $content-width;
			display: flex;
			align-items: flex-start;
			justify-content: center;

			&__no-auth {
				width: 100%;
			}

			&__content {
				flex: 1;
				margin-top: 20px;
			}
		}
	}
}
