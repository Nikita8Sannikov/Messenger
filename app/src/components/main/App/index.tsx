import Router from "@components/main/Router";
import ServiceComponents from "@components/main/ServiceComponents";
import HeaderComponent from "@components/layouts/header";
import MenuComponent from "@components/layouts/menu";
import SpinnerComponent from "@components/ui/spinner";
import { selectMainState } from "@store/main/slice";
import { useAppSelector } from "@hooks/useGlobalState";

import "./app.scss";

// Главный компонент, который отрисовывает основную верстку проекта
export default function App() {
	const { isAuth, loading } = useAppSelector(selectMainState);
	
	return <div className="root">
		<ServiceComponents />

		{loading
			? <SpinnerComponent />
			: <>
				{isAuth ? <HeaderComponent /> : null}

				<div className="root__wrapper">
					{/* <button onClick={() => setIsDarkMode((prev: boolean) => !prev)}>Toggle Theme</button> */}
					<div
						className={`root__wrapper__container ${
							isAuth
								? ""
								: "root__wrapper__container__no-auth"
						}`}
					>
						{isAuth ? <MenuComponent /> : null}

						<div className="root__wrapper__container__content">
							<Router />
						</div>
					</div>
				</div>
			</>
		}
	</div>
}
