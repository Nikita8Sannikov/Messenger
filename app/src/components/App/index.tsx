import React from "react";

import { selectMainState } from "../../state/main/slice";
import { useAppSelector } from "../../hooks/useGlobalState";
import Router from "../../service/Router";
import ServiceComponents from "../../service/ServiceComponents";
import Header from "../Header";
import MenuComponent from "../Menu";

import "./app.scss";

export default function App() {
  const { isAuth, loading } = useAppSelector(selectMainState);

  return <div className="root">
    <ServiceComponents />

    {loading
      ? null
      : <>
        {isAuth ? <Header /> : null}

        <div className="root__wrapper">
          <div className={`root__wrapper__container ${isAuth ? "" : "root__wrapper__container__no-auth"}`}>
            {isAuth ? <MenuComponent /> : null}

            <div className="root__wrapper__container__content">
              <Router />
            </div>
          </div>
        </div>
      </>
    }
  </div>
};