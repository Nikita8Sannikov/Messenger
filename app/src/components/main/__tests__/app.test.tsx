import { render } from "@testing-library/react";

import App from "@components/main/app";
import useUserStore, { type UserStore } from "../../../__mocks__/@store/user";

type MockUserStore = jest.Mock & {
    getState: () => UserStore;
};

describe("App Component", () => {
	const mockUser = {
		id: 1,
		name: "Test User",
		myAvatar: {
			userId: "123",
			src: "avatar.jpg",
			alt: "User Avatar",
		},
	};

	beforeEach(() => {
		jest.clearAllMocks();
		const mockStore = {
			isUserLoading: false,
			user: null,
			userDetails: null,
			editUserDetails: null,
			myAvatar: null,
			setUser: jest.fn(),
			setLoadingUser: jest.fn(),
			setUserDetails: jest.fn(),
			setMyAvatar: jest.fn(),
			reset: jest.fn(),
			getState: () => mockStore,
		};
		const mockUseUserStore = useUserStore as unknown as MockUserStore;
		mockUseUserStore.mockImplementation((selector) => selector(mockStore));
		mockUseUserStore.getState = () => mockStore;
	});

	it("renders loading spinner when user is loading", () => {
		const mockStore = {
			isUserLoading: true,
			user: null,
			userDetails: null,
			editUserDetails: null,
			myAvatar: null,
			setUser: jest.fn(),
			setLoadingUser: jest.fn(),
			setUserDetails: jest.fn(),
			setMyAvatar: jest.fn(),
			reset: jest.fn(),
			getState: () => mockStore,
		};
		const mockUseUserStore = useUserStore as unknown as MockUserStore;
		mockUseUserStore.mockImplementation((selector) => selector(mockStore));
		mockUseUserStore.getState = () => mockStore;

		const { container } = render(<App />);
		expect(document.querySelector(".spinner")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("renders authenticated layout when user is loaded and authenticated", () => {
		const mockStore = {
			isUserLoading: false,
			user: mockUser,
			userDetails: null,
			editUserDetails: null,
			myAvatar: mockUser.myAvatar,
			setUser: jest.fn(),
			setLoadingUser: jest.fn(),
			setUserDetails: jest.fn(),
			setMyAvatar: jest.fn(),
			reset: jest.fn(),
			getState: () => mockStore,
		} as unknown as UserStore;
		const mockUseUserStore = useUserStore as unknown as MockUserStore;
		mockUseUserStore.mockImplementation((selector) => selector(mockStore));
		mockUseUserStore.getState = () => mockStore;

		const { container } = render(<App />);

		expect(container.querySelector(".header")).toBeInTheDocument();
		expect(container.querySelector(".menu")).toBeInTheDocument();
		expect(container.querySelector(".main-content")).toBeInTheDocument();

		expect(container.querySelector(".root__wrapper")).not.toHaveClass("root__no-auth");
		expect(container).toMatchSnapshot();
	});

	it("renders unauthenticated layout when user is loaded but not authenticated", () => {
		const mockStore = {
			isUserLoading: false,
			user: null,
			userDetails: null,
			editUserDetails: null,
			myAvatar: null,
			setUser: jest.fn(),
			setLoadingUser: jest.fn(),
			setUserDetails: jest.fn(),
			setMyAvatar: jest.fn(),
			reset: jest.fn(),
			getState: () => mockStore,
		};
		const mockUseUserStore = useUserStore as unknown as MockUserStore;
		mockUseUserStore.mockImplementation((selector) => selector(mockStore));
		mockUseUserStore.getState = () => mockStore;

		const { container } = render(<App />);

		expect(container.querySelector(".header")).not.toBeInTheDocument();
		expect(container.querySelector(".menu")).not.toBeInTheDocument();

		expect(container.querySelector(".main-content")).toBeInTheDocument();
		expect(container.querySelector(".root__wrapper")).toHaveClass("root__no-auth");
		expect(container).toMatchSnapshot();
	});
});
