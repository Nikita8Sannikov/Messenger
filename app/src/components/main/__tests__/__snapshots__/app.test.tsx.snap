// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders authenticated layout when user is loaded and authenticated 1`] = `
<div>
  <div
    class="root"
  >
    <header
      class="header"
      data-testid="header"
    >
      <div
        class="header-container"
      >
        <div
          class="header-container__toolbar"
        >
          <div
            class="header-container__toolbar__logo"
          >
            <svg
              data-testid="main-logo"
              fill="none"
              height="24"
              width="136"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                clip-path="url(#new_logo_vk_with_text__a)"
              >
                <g
                  clip-path="url(#new_logo_vk_with_text__b)"
                >
                  <path
                    clip-rule="evenodd"
                    d="M67 12.5c0 3.34-2.43 5.5-5.88 5.5-3.45 0-5.88-2.16-5.88-5.5S57.67 7 61.12 7C64.57 7 67 9.16 67 12.5Zm-9.22 0c0 2.07 1.35 3.5 3.34 3.5s3.34-1.43 3.34-3.5-1.35-3.45-3.34-3.45-3.34 1.38-3.34 3.45Zm-17.03-.21c.95-.44 1.56-1.18 1.56-2.33 0-1.73-1.58-2.96-3.87-2.96h-5.27v11h5.5c2.37 0 4.02-1.29 4.02-3.05 0-1.33-.87-2.32-1.94-2.66ZM35.6 9.01h2.83c.85 0 1.44.5 1.44 1.2s-.6 1.2-1.44 1.2h-2.83V9ZM38.67 16h-3.06V13.3h3.06c.96 0 1.59.55 1.59 1.36 0 .8-.63 1.33-1.59 1.33ZM51.84 18h3.19l-5.06-5.71L54.61 7h-2.9l-3.68 4.27h-.6V7H45v11h2.44v-4.38h.59l3.8 4.38ZM76.47 7v4.34h-4.93V7H69.1v11h2.43v-4.44h4.93V18h2.43V7h-2.43ZM86.9 18h-2.44V9.22h-3.8V7H90.7v2.22h-3.8V18Zm9.7-11c-2.14 0-4.02.89-4.57 2.8l2.24.37a2.38 2.38 0 0 1 2.2-1.25c1.33 0 2.12.9 2.22 2.33h-2.37c-3.23 0-4.84 1.42-4.84 3.45 0 2.05 1.59 3.3 3.83 3.3 1.8 0 3-.82 3.53-1.73l.5 1.73h1.8v-6.18c0-3.19-1.73-4.82-4.54-4.82Zm-.72 9.16c-1.19 0-1.95-.61-1.95-1.57 0-.84.62-1.43 2.48-1.43h2.3c0 1.8-1.14 3-2.83 3ZM113.73 18h-3.2l-3.8-4.38h-.6V18h-2.42V7h2.43v4.27h.59L110.4 7h2.9l-4.63 5.29 5.05 5.71Zm4.27 0h2.44V9.22h3.8V7H114.2v2.22h3.8V18Zm12.3-11c3.33 0 5.7 2.2 5.7 5.37 0 .3-.02.55-.04.79h-8.84c.23 1.69 1.46 2.83 3.32 2.83 1.29 0 2.3-.55 2.83-1.33l2.29.38c-.83 2.1-2.98 2.96-5.27 2.96-3.34 0-5.71-2.18-5.71-5.5s2.37-5.5 5.71-5.5Zm3.06 4.25A3.06 3.06 0 0 0 130.29 9a3 3 0 0 0-3.02 2.25h6.09Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                  <path
                    d="M11.5 24h1c5.44 0 8.15 0 9.83-1.68C24 20.64 24 17.92 24 12.5v-1.02c0-5.4 0-8.12-1.67-9.8C20.65 0 17.93 0 12.5 0h-1C6.06 0 3.35 0 1.67 1.68 0 3.36 0 6.08 0 11.5v1.02c0 5.4 0 8.12 1.68 9.8C3.36 24 6.08 24 11.5 24Z"
                    fill="#07F"
                  />
                  <path
                    d="M12.77 17.29c-5.47 0-8.59-3.75-8.72-9.99h2.74c.09 4.58 2.11 6.52 3.71 6.92V7.3h2.58v3.95c1.58-.17 3.24-1.97 3.8-3.95h2.58a7.62 7.62 0 0 1-3.51 4.98 7.9 7.9 0 0 1 4.11 5.01h-2.84a4.94 4.94 0 0 0-4.14-3.57v3.57h-.31Z"
                    fill="#fff"
                  />
                </g>
              </g>
            </svg>
          </div>
          <div
            class="header-container__toolbar__avatar"
          >
            <div>
              <span
                class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular avatar user-avatar css-fmeqll-MuiAvatar-root"
                  id="user-avatar"
                >
                  <img
                    alt="User Avatar"
                    class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                    src="avatar.jpg"
                  />
                </div>
                <span
                  class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      class="root__wrapper "
    >
      <div
        class="root__wrapper__container"
      >
        <div
          class="menu"
          data-testid="menu"
        >
          <nav
            class="menu__nav"
          >
            <ul
              class="MuiList-root MuiList-padding menu__list undefined css-b7dj7n-MuiList-root"
              role="menu"
              tabindex="-1"
            >
              <li
                class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters menu__nav__item css-1rju2q6-MuiButtonBase-root-MuiMenuItem-root"
                data-testid="menu-item"
                role="menuitem"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-1wtp6o0-MuiSvgIcon-root"
                  data-testid="AccountCircleOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2M7.35 18.5C8.66 17.56 10.26 17 12 17s3.34.56 4.65 1.5c-1.31.94-2.91 1.5-4.65 1.5s-3.34-.56-4.65-1.5m10.79-1.38C16.45 15.8 14.32 15 12 15s-4.45.8-6.14 2.12C4.7 15.73 4 13.95 4 12c0-4.42 3.58-8 8-8s8 3.58 8 8c0 1.95-.7 3.73-1.86 5.12"
                  />
                  <path
                    d="M12 6c-1.93 0-3.5 1.57-3.5 3.5S10.07 13 12 13s3.5-1.57 3.5-3.5S13.93 6 12 6m0 5c-.83 0-1.5-.67-1.5-1.5S11.17 8 12 8s1.5.67 1.5 1.5S12.83 11 12 11"
                  />
                </svg>
                <div
                  class="menu__nav__item__title"
                >
                  My profile
                </div>
              </li>
              <li
                class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters menu__nav__item css-1rju2q6-MuiButtonBase-root-MuiMenuItem-root"
                data-testid="menu-item"
                role="menuitem"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-1wtp6o0-MuiSvgIcon-root"
                  data-testid="MessageOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M4 4h16v12H5.17L4 17.17zm0-2c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm2 10h12v2H6zm0-3h12v2H6zm0-3h12v2H6z"
                  />
                </svg>
                <div
                  class="menu__nav__item__title"
                >
                  Messanger
                </div>
                <span
                  class="MuiBadge-root notification-badge css-63ezxx-MuiBadge-root"
                  data-testid="notification-badge"
                >
                  <span
                    class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular css-2ao5rv-MuiBadge-badge"
                  >
                    0
                  </span>
                </span>
              </li>
              <li
                class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters menu__nav__item css-1rju2q6-MuiButtonBase-root-MuiMenuItem-root"
                data-testid="menu-item"
                role="menuitem"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-1wtp6o0-MuiSvgIcon-root"
                  data-testid="PeopleOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M9 13.75c-2.34 0-7 1.17-7 3.5V19h14v-1.75c0-2.33-4.66-3.5-7-3.5M4.34 17c.84-.58 2.87-1.25 4.66-1.25s3.82.67 4.66 1.25zM9 12c1.93 0 3.5-1.57 3.5-3.5S10.93 5 9 5 5.5 6.57 5.5 8.5 7.07 12 9 12m0-5c.83 0 1.5.67 1.5 1.5S9.83 10 9 10s-1.5-.67-1.5-1.5S8.17 7 9 7m7.04 6.81c1.16.84 1.96 1.96 1.96 3.44V19h4v-1.75c0-2.02-3.5-3.17-5.96-3.44M15 12c1.93 0 3.5-1.57 3.5-3.5S16.93 5 15 5c-.54 0-1.04.13-1.5.35.63.89 1 1.98 1 3.15s-.37 2.26-1 3.15c.46.22.96.35 1.5.35"
                  />
                </svg>
                <div
                  class="menu__nav__item__title"
                >
                  Friends
                </div>
                <span
                  class="MuiBadge-root notification-badge css-63ezxx-MuiBadge-root"
                  data-testid="notification-badge"
                >
                  <span
                    class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular css-2ao5rv-MuiBadge-badge"
                  >
                    0
                  </span>
                </span>
              </li>
              <li
                class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters menu__nav__item css-1rju2q6-MuiButtonBase-root-MuiMenuItem-root"
                data-testid="menu-item"
                role="menuitem"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-1wtp6o0-MuiSvgIcon-root"
                  data-testid="CameraAltIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="3.2"
                  />
                  <path
                    d="M9 2 7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5"
                  />
                </svg>
                <div
                  class="menu__nav__item__title"
                >
                  Photos
                </div>
              </li>
            </ul>
          </nav>
          <div
            class="menu__down-info"
          >
            To developers
          </div>
        </div>
        <div
          class="main-content MuiBox-root css-0"
          data-testid="box-component"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`App Component renders loading spinner when user is loading 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="spinner"
      data-testid="spinner"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`App Component renders unauthenticated layout when user is loaded but not authenticated 1`] = `
<div>
  <div
    class="root"
  >
    <div
      class="root__wrapper root__no-auth"
    >
      <div
        class="root__wrapper__container"
      >
        <div
          class="main-content MuiBox-root css-0"
          data-testid="box-component"
        />
      </div>
    </div>
  </div>
</div>
`;
